crates/red_knot_python_semantic/src/types.rs

```diff
@@ -463,6 +463,52 @@ impl<'db> Type<'db> {
         self == other
     }
 
+    /// Return true if there is just a single inhabitant for this type.
+    ///
+    /// Note: This function aims to have no false positives, but might return `false`
+    /// for more complicated types that are actually singletons.
+    pub(crate) fn is_singleton(self, db: &'db dyn Db) -> bool {
+        match self {
+            Type::Any
+            | Type::Never
+            | Type::Unknown
+            | Type::Todo
+            | Type::Unbound
+            | Type::Module(..)
+            | Type::Instance(..)
+            | Type::IntLiteral(..)
+            | Type::StringLiteral(..)
+            | Type::BytesLiteral(..)
+            | Type::LiteralString => {
+                // Note: The literal types included in this pattern are not true singletons.
+                // There can be multiple Python objects (at different memory locations) that
+                // are both of type Literal[345], for example.
+                false
+            }
+            Type::None | Type::BooleanLiteral(_) | Type::Function(..) | Type::Class(..) => true,
+            Type::Tuple(tuple) => tuple.elements(db).iter().all(|ty| ty.is_singleton(db)),
```

Comment:
As I mentioned in our call just now, this unfortunately isn't correct. Two tuples are not guaranteed to be the same object in terms of _identity_ (which literally refers to the memory address of the underlying object) even if they compare equal, and even if all of their elements are in fact references to the exact same singletons:

```pycon
>>> x = (False, True)
>>> y = (False, True)
>>> False is False
True
>>> True is True
True
>>> all(x_element is y_element for x_element, y_element in zip(x, y))
True
>>> all(id(x_element) == id(y_element) for x_element, y_element in zip(x, y))
True
>>> id(x)
4378553856
>>> id(y)
4378331520
>>> x is y
False
```

The one exception is the empty tuple, which _is_ always the same object:

```pycon
>>> x = ()
>>> y = ()
>>> id(x)
4386737520
>>> id(y)
4386737520
>>> x is y
True
```

...However... I _thought_ this was guaranteed by the language reference... but now that I check, it appears the language reference [says the opposite](https://docs.python.org/3/reference/expressions.html#parenthesized-forms):

> An empty pair of parentheses yields an empty tuple object. Since tuples are immutable, the same rules as for literals apply (i.e., two occurrences of the empty tuple may or may not yield the same object).

So I'm not sure exactly what we should do here. The empty tuple is commonly used as a sentinel value, so I highly doubt that Python will realistically ever change it so that it is no longer a singleton; the backwards-compatibility implications would be too great, even if this is not guaranteed by the language spec. Perhaps the best course of action would be to treat it as a singleton in our model, but link to the that section of the language spec and say that this isn't strictly guaranteed by the language.

Reply:
> Perhaps the best course of action would be to treat it as a singleton in our model, but link to the that section of the language spec and say that this isn't strictly guaranteed by the language.

I did that for now.

Reply:
I've started a discussion on Discourse about whether it's time to change the language reference so that this is a guaranteed property of the language: https://discuss.python.org/t/should-we-specify-in-the-language-reference-that-the-empty-tuple-is-a-singleton/67957/1